File: COINS.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"

long long solve(long long n, long long dp[]) {
	if (n < 12)
		return n;
	if (dp[n])
		return dp[n];
	return dp[n] = max(n, solve(n/2, dp) + solve(n/3, dp) + solve(n/4, dp));
}

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
#endif

	long long n;
	while (cin >> n) {
		long long *dp = new long long[n+1];
		cout << solve(n, dp) << endl;
	}
}

File: COINS.py
def solve(n):
	if n < 12:
		return n
	if n in dp.keys():
		return dp[n]
	dp[n] = max(n, solve(n//2) + solve(n//3) + solve(n//4))
	return dp[n]

while True:
	try:
		n = int(input())
	except:
		break
	dp = {}
	print(solve(n))
